chatgpt-shell.el

```diff
@@ -608,6 +608,53 @@ non-nil; otherwise `completing-read'."
         (setq-default chatgpt-shell-model-version selection))
     (error "No other providers found")))
 
+(defun chatgpt-shell-unsorted-collection (collection)
+  "Return a completion table based on COLLECTION that inhibits sorting."
+  (lambda (string predicate action)
+    (if (eq action 'metadata)
+        (let ((current-metadata (cdr (completion-metadata (minibuffer-contents)
+                                                          collection
+                                                          minibuffer-completion-predicate))))
+          `(metadata
+            ,@(map-merge 'alist
+                         current-metadata
+                         '((display-sort-function . identity)
+                           (cycle-sort-function . identity)))))
+      (complete-with-action action collection string predicate))))
+
+(defun chatgpt-shell-select-reasoning-effort (&optional global)
+  "Interactively set the reasoning effort for the current model.
+
+By default, this is done buffer-locally when in a
+`chatgpt-shell-mode' buffer `chatgpt-shell-prompt-compose-mode'
+buffer. When GLOBAL is non-nil (interactively with a prefix
+argument), it is set globally."
+  (interactive "P")
+  (let* ((model (chatgpt-shell--resolved-model))
+         (selector (map-elt model :reasoning-effort-selector)))
+    (unless selector
+      (user-error "No reasoning effort selector is defined for %s" (chatgpt-shell-model-version)))
+    (let* ((first t)
+           (buf (cond
+                 (global
+                  nil)
+                 ((eq major-mode 'chatgpt-shell-mode)
+                  (current-buffer))
+                 ((memq major-mode '(chatgpt-shell-prompt-compose-mode chatgpt-shell-prompt-compose-view-mode))
+                  (chatgpt-shell--primary-buffer))))
+           (result (with-current-buffer buf (funcall selector model))))
+      (dolist (cell result)
+        (let* ((sym (car cell))
+               (var (if buf
+                        (with-current-buffer buf
+                          (make-local-variable sym))
+                      sym))
+               (val (cdr cell)))
+          (set var val)
```

Comment:
I'm guessing this is creating a variable that can be looked up using `describe-variable`? Where is the variables declared?

Reply:
No, these variables are all already declared. Calling `make-local-variable` is a standard method used in elisp code to create a value that causes it to be set buffer locally when `set` is used. I'll add a comment.

Reply:
Thanks for this.

> No, these variables are all already declared

I missed them. Could you share the name of one as an example, so I can search for them? I would also add this as a comment in code.

Reply:
You're right. I missed the vars in the example. Super handy to see them in the comment as anyone can read the comment and know what to search for to find related code. Similarly, using an alist as proposed in my previous comment also strengthens this searchable aspect.